data_utils/data.py

```diff
@@ -280,10 +282,7 @@ def reader():
             lambda instance: self.process_utterance(instance["audio_filepath"], instance["text"]),
             reader, self._num_threads, 4096)
 
-        # register callback to main process
-        atexit.register(cleanup_callback)
-
-        return reader
+        return reader, cleanup_callback
```

Comment:
Here, we register `cleanup_callback` to atexit to handle the interruption of main process. So, I think better to keep the logic here.

Reply:
We should do cleanup at every epoch end, not at main process exit.
Every epoch did create a batch of workers, and would clean up them atexit. That's the cause of the leak.

Reply:
Would use `try...finally` be better?

Reply:
I know what you mean, however, I think registering to atexit and cleaning up after each epoch can coexist.

Reply:
I think you didn't understand what i mean. Sorry for my poor English.
I mean 'register so many callbacks atexit' is a kind of leak. So, maybe use `try...finally` is better?

Reply:
Seems `try...finally` can't catch the signal like SIGTERM or SIGKILL. I agree with you, however we can clean up after each epoch and make sure only one clean callback registered to atexit?

Reply:
Infact, atexit also cannot handle signal SIGTERM and SIGKILL.
Should we care about those signals?